docs/example_java.md

```diff
@@ -223,47 +223,129 @@ Let's see what this looks like!
+If we unpacked the fat Jar, we would find that the Jar contains many files
+internally. These files naturally lend themselves to Docker layers. 
+Java Jars were using layer caches before Docker made them cool. How can we take
+advantage of this?
+
+We've updated [our
+Tiltfile](https://github.com/windmilleng/tilt-example-java/blob/master/3-unpacked/Tiltfile)
+to unpack the Jar in the `build/jar` directory:
+
+```python
+local_resource(
+  'example-java-compile',
+  './gradlew bootJar && ' +
+  'unzip -o build/libs/example-0.0.1-SNAPSHOT.jar -d build/jar',
+  deps=['src', 'build.gradle'],
+  resource_deps = ['deploy'])
+```
+
+We've also updated our [Dockerfile](https://github.com/windmilleng/tilt-example-java/blob/master/3-unpacked/Dockerfile):
+
+```
+FROM openjdk:8-jre-alpine
+
+WORKDIR /app
+ADD BOOT-INF/lib /app/lib
+ADD META-INF /app/META-INF
+ADD BOOT-INF/classes /app
+
+ENTRYPOINT java -cp .:./lib/* dev.tilt.example.ExampleApplication
+```
+
+This Dockerfile adds files from `build/jar` in order from least frequently
+used to most frequently used, to improve caching.
+
+The Dockerfile also has a new the entrypoint to load the main application class,
+since we're no longer using an executable Jar.
+
+Let's see what this looks like!
+
+<figure>
+  <a class="is-image" href="https://cloud.tilt.dev/snapshot/AcaQnucLOBUS4TGQauw=">
+    <img src="assets/docimg/example-java-image-4.png">
+  </a>
+  <figcaption>Step 3 complete. Click the screenshot to see an interactive snapshot.</figcaption>
+</figure>
+
+| Approach | Deploy Time |
+|---|---|
+| Naive | 87.7s |
+| Local Compile | 13.4s |
+| Optimized Dockerfile | 6.5s |
+{:.benchmark-report}
+
+If you don't want to optimize the Dockerfile yourself,
+check out [Jib](https://github.com/GoogleContainerTools/jib)!
+
+Jib is a Java image builder that re-packs Java Jars as container images using
+similar tricks. There are Jib plugins for Maven and Gradle.  The
+[tilt-example-java](https://github.com/windmilleng/tilt-example-java) repo has
+an example
+[Tiltfile](https://github.com/windmilleng/tilt-example-java/blob/master/101-jib/Tiltfile)
+that uses `custom_build` to generate images with Jib.
+
+## Step 4: Let's Live Update It
+
+When we make a change to a file, we currently have to build an image, deploy new
+Kubernetes configs, and wait for Kubernetes to schedule the pod.
 
 With Tilt, we can skip all of these steps, and instead
 [live_update](live_update_tutorial.html) the pod in place.
 
-Here's our [new Tiltfile](https://github.com/windmilleng/tilt-example-java/blob/master/3-recommended/Tiltfile) 
+Here's our [new Tiltfile](https://github.com/windmilleng/tilt-example-java/blob/master/4-recommended/Tiltfile) 
 with the following new code:
 
 ```python
+local_resource(
+  'example-java-compile',
+  './gradlew bootJar && ' +
+  'unzip -o build/libs/example-0.0.1-SNAPSHOT.jar -d build/jar-staging && ' +
+  'rsync --inplace --checksum -r build/jar-staging/ build/jar',
```

Comment:
Clever! 👏

(Though I wonder if the added indirection of needing a staging folder won't make people confused.)

Reply:
ya, i wonder if live_update should have some sort of checksum feature, so that you don't need to do this.